pkg/estimator/pb/types.go

```diff
@@ -57,6 +57,10 @@ type ReplicaRequirements struct {
 	// ResourceRequest represents the resources required by each replica.
 	// +optional
 	ResourceRequest corev1.ResourceList `json:"resourceRequest,omitempty" protobuf:"bytes,2,rep,name=resourceRequest,casttype=k8s.io/api/core/v1.ResourceList,castkey=k8s.io/api/core/v1.ResourceName"`
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,3,opt,name=namespace"`
```

Comment:
I make it optional because we want to backward compatibility ?

Reply:
Does the backward compatibility mean a lower version(like v1.8) of `karmada-scheduler-estimator` which can not handle namespace`, but it can work with a higher version (like the coming v1.9) of `karmada-scheduler`?
When the `karmada-scheduler-estimator` receives a request, it will drop the fields that it doesn't recognized. am I right?

Reply:
> I make it optional because we want to backward compatibility ?

Required field could be backward compatible as well, all the non-omitempty fields that are not filled will be empty strings. BTW, we have already transfered name and namespace between scheduler and estimator, you can pay attention to the input string `object`.

``` go
func (es *AccurateSchedulerEstimatorServer) EstimateReplicas(ctx context.Context, object string, request *pb.MaxAvailableReplicasRequest) (int32, error) 
```

Reply:
> all the non-omitempty fields that are not filled will be empty strings. 

I'm not familiar with this. I'm not sure if I understand correctly.  Both client and server must handle the `required` field, right? So, client and server must upgrade together, isn't?

Seems the `required` field has been deprecated from protoc3.

Echo from https://protobuf.dev/programming-guides/proto2/#field-labels:
> required: Do not use. Required fields are so problematic they were removed from proto3. Semantics for required field should be implemented at the application layer. When it is used, a well-formed message must have exactly one of this field.

Reply:
> Both client and server must handle the required field, right? So, client and server must upgrade together, isn't?

Well, this `required` that we are talking about is not about required fields in pb. I mean something like this:

```go
        // +required
	Namespace string `json:"namespace" protobuf:"bytes,3,opt,name=namespace"`
```
So it only affects the json format. We use go-to-protobuf to generate proto files, all fields in that file is `optional` or `repeated`, so it does not matter.

Reply:
I'm trying to understand what you mean:

Presiously, the generated `.proto` file is

```
message ReplicaRequirements {
  // +optional
  optional string namespace = 3;
}
```

after introduced your `// +required` and removed `omitempty `, the generated `.proto` file will become like this

```
message ReplicaRequirements {
  // +required
  optional string namespace = 3;
}
```

This comment is just a way of annotating the fields, they do not actually affect the runtime behaviour of gRPC. 

At runtime of gRPC, they are all optional fields and are backward compatible.

Or put it more directly, the two are simply different annotations for fields. It won't actually work, it's just for more readable (If you need to ensure that a field is required or optional, you ultimately rely on validating and processing it in the other custom program logic).

Am I right?

Reply:
Exactly.